MYBackgroundMonitor.m

```diff
@@ -22,6 +22,16 @@ @implementation MYBackgroundMonitor
 @synthesize onBackgroundTaskExpired=_onBackgroundTaskExpired;
 
 
+static BOOL runningInAppExtension() {
+    return [[[[NSBundle mainBundle] bundlePath] pathExtension] isEqualToString: @"appex"];
+}
+
+
+static UIApplication* sharedApplication() {
+    return [[UIApplication class] performSelector: @selector(sharedApplication)];
```

Comment:
If the goal is to avoid linking against UIKit, this won't work. Calling `[UIApplication class]` creates an reference to the class symbol. To avoid that you have to use `NSClassFromString(@"UIApplication")` instead. And you don't need `performSelector:`; you can just call `sharedApplication` normally.

Reply:
Also, if the `NSClassFromString` call returns nil, you can assume you are running in a context where there is no UIKit or UIApplication, e.g. in an app extension. That seems like a more reliable test than checking the app bundle's filename extension.

Reply:
I have checked the behavior what a test app extension app. UIKit and UIApplication are available for App Extension. Only `sharedApplication` is not permitted to use. XCode shows the following error when using sharedApplication method  

```
'sharedApplication' is unavailable: not available on iOS (App Extension) - Use view controller based solutions where appropriate instead.
```